### ISSUE TITLE
Bundled interaction not taken into account as soon as the confirmation from the Sequencer is received

### ISSUE THREAD

**Issue by noomly** [redstone-finance/redstone-smartcontracts#137]
Sometimes, after running a first `bundleInteraction`, running a second one right after that depends on it throws with `Error: Cannot create interaction: [...]` as if the first interaction hasn't been taken into account yet by the Sequencer. If I try to run the second interaction again after a few seconds it works.

Note: I do wait for the confirmation from the Sequencer for every interactions.

**Comment by ppedziwiatr** [comment]
@noomly , do you happen to have any MRE? this would greatly help in the analysis...

**Comment by noomly** [comment]
I am currently working on it yes. I will link a repo here later.

**Comment by noomly** [comment]
Here is the repo: https://github.com/pianity/bundle-interaction-issue

I initially wrote a very simple test case using a contract as simple as possible, however it turned out that I couldn't reproduce the issue with it... I still included it to the repo and added my original test case using an ERC1155 inspired contract, for which the issue is easily reproducible. I still think the issue shouldn't come from the contract as it doesn't matter whether an interaction is executed immediately after another or not from the contract's point of view. This could maybe give a clue as to where the problem is? The difference could lie in the execution time of the contract... although I'm not sure how it could cause this problem.

**Comment by ppedziwiatr** [comment]
Thanks @noomly , I'll analyse this by the end of this week.

**Comment by noomly** [comment]
`init` and `transfer` both return a `Promise` so there is no actual need to await transfer here as the result of `init` is awaited here https://github.com/ppedziwiatr/bundle-interaction-issue/blob/main/src/pianityContractTest.ts#L62 anyway

**Comment by ppedziwiatr** [comment]
From Discord...


So there are several aspects:
1. the sequencer is using the arweave.net/info to load the current block height - it is being used to generate part of the sequence.
But the arweave.net/info underneath proxies the request to one of the gateway's backing nodes - each of them might be synchronized at a slightly different block height (the difference is usually max one block) - and return different value of the current network block height at a given time.

2. the sdk - if the block height parameter is not passed to the readState method - will also ask arweave.net/info for the current block height and evaluate the state up to that block height

It may happen (because of the fact that the backing nodes might be synchronized at a different block heights) that the lastly posted interaction will have sequence generated with block height ONE higher, than the block height that the SDK is using to load the latest interactions.

An example:
The first screenshot shows that transfered to account B PQmYeljLIN_bPHP_PHUJpn93FEmc36kq7_RNBSdk6v0 was sucessfull.
![image](https://user-images.githubusercontent.com/84311757/166951292-aefde03b-d454-4a14-aa10-9074f4cab050.png)

After this it tries to load the interactions up to 926636 height (because that's what arweave.net/info returned at the moment the SDK was asking for the current network block height).

BUT
on the sequencer/gateway side the PQmYeljLIN_bPHP_PHUJpn93FEmc36kq7_RNBSdk6v0 transaction has been generated with 926637 height (2nd. screenshot) - because that was the height that the arweave.net/info returned at the time the sequencer was asking for the current network height.
![image](https://user-images.githubusercontent.com/84311757/166951436-99e65c91-5ec8-49f0-8c04-610f8f9a6c0c.png)


Because of that and because SDK was asking for interactions up to  926636 height - the lastly created interaction PQmYeljLIN_bPHP_PHUJpn93FEmc36kq7_RNBSdk6v0 was not returned - and the dryRun (which is run in the strict mode used by this test) failed.

tl;dr
The core issue is that arweave.net/info might not return "stable" network height results and the SDK and sequencer might use slightly different heights at any given point of time.

**Comment by ppedziwiatr** [comment]
In order to fix the issue:
1. Redstone Gateway now supplies a new endpoint - https://gateway.redstone.finance/gateway/arweave/info - that is an equivalent of https://arweave.net/info, but guarantees that block heights will be always returned in an ascending order
2. The SDK is using by default the above endpoint to load current network height - if it is operating on RedStone Gateway (i.e. `useRedstoneGateway()` method is called during sdk instance creation)
3. some other minor improvements - e.g. sdk is now forced to cache every 100 interactions. In case of interactions from sequencer, the SDK has a built in protection against caching at the "last" block height (as new transactions from Sequencer might appear on this block height after the state was cached by the SDK).
But - in case the SDK is caching only after full state evaluation (currently default mode) - it might have happened that no state was cached at all - hence the requirement to cache every 100 interactions (which should not impact the performance).
